source/cppfront.cpp

```diff
@@ -1702,32 +1762,38 @@ class cppfront
-                            last_postfix_expr_was_pointer = true;
-                        }
-                        else
-                        {
-                            if (n.ops.front().op->type() == lexeme::PlusPlus ||
-                                n.ops.front().op->type() == lexeme::MinusMinus ||
-                                n.ops.front().op->type() == lexeme::LeftBracket
-                                ) {
-                                errors.emplace_back(
-                                    n.ops.front().op->position(),
-                                    n.ops.front().op->to_string(true) + " - pointer arithmetic is illegal - use std::span or gsl::span instead"
-                                );
-                                violates_bounds_safety = true;
-                            }
-                            else if (n.ops.front().op->type() == lexeme::Tilde) {
-                                errors.emplace_back(
-                                    n.ops.front().op->position(),
-                                    n.ops.front().op->to_string(true) + " - pointer bitwise manipulation is illegal - use std::bit_cast to convert to raw bytes first"
-                                );
+                //  TODO: Generalize this: 
+                //        - we don't recognize pointer types from Cpp1
+                //        - we don't deduce pointer types from parameter_declaration_list_node
+                if ( is_pointer_declaration(unqual->identifier) ) {
+                    if (n.ops.empty()) {
+                        last_postfix_expr_was_pointer = true;
+                    }
+                    else
+                    {
+                        auto op = [&](){
+                            if (n.ops.size() >= 2 && n.ops[0].op->type() == lexeme::LeftParen) {
+                                return n.ops[1].op;
+                            } else {
+                                return n.ops.front().op;
                             }
+                        }();
+
+                        if (op->type() == lexeme::PlusPlus ||
+                            op->type() == lexeme::MinusMinus ||
+                            op->type() == lexeme::LeftBracket
+                            ) {
+                            errors.emplace_back(
+                                op->position(),
+                                op->to_string(true) + " - pointer arithmetic is illegal - use std::span or gsl::span instead"
+                            );
+                            violates_bounds_safety = true;
+                        }
+                        else if (op->type() == lexeme::Tilde) {
+                            errors.emplace_back(
+                                op->position(),
+                                op->to_string(true) + " - pointer bitwise manipulation is illegal - use std::bit_cast to convert to raw bytes first"
+                            );
```

Comment:
I haven't had the cycles to really grok this, and the only reason I'm reading through the PR is because (a) the test case at top is clear and (b) the code change is mostly localized here. That limits the downside because it should be very local.

I'm not yet sure whether I should merge this... In the long term I aim to implement my C++ Core Guidelines Lifetime analysis, and that will likely replace at least parts of this work. But if we do merge this now, I still view it as an interim '[expanded] initial partial implementation' and don't think we would want to go further in this direction in the near term. First we would want to implement the Lifetime analysis and then see where we are, and the Lifetime implementation itself needs to be farther in the future after classes and metaclass functions.

Reply:
The main change here is about calling a new function `is_pointer_declaration(unqual->identifier)` and to check not only the operator that is called on the node but also the results of the function call (e.g. `n()++`)